File: kernel/src/heap.zig
//! Kernel heap allocation and management.
//! The allocator is currently a very simple linked list allocator, although this
//! may change in future.

const std = @import("std");
const Allocator = std.mem.Allocator;
const Error = Allocator.Error;
const builtin = @import("builtin");
const root = @import("root");
const compileErrorFmt = root.zig_extensions.compileErrorFmt;
const smp = root.smp;
const arch = root.arch;
const page_allocator = arch.page_allocation.page_allocator_ptr;
const alignToPage = arch.paging.alignToPage;

const logger = std.log.scoped(.heap);

pub const Block = packed struct {
    /// Length of the block area, should be u30 for 32-bit and u62 for 64-bit
    len: LenType,
    /// Whether the block is used, if `false` the block is free
    used: bool,
    /// Whether there is another block following this block
    next: bool,

    pub const LenType = @Type(.{.Int = .{
        .is_signed = false,
        .bits = @bitSizeOf(usize) - 2,
    }});

    pub const alignment: u29 = @alignOf(usize);

    pub fn init(len: usize, used: bool, next: bool) Block {
        return Block{
            .len = @truncate(LenType, len),
            .used = used,
            .next = next,
        };
    }

    pub fn getStartPtr(self: *Block) [*]u8 {
        return @ptrCast([*]u8, self) + @sizeOf(Block);
    }

    pub fn getNext(self: *Block) ?*Block {
        if (!self.next) return null;
        const address = @ptrToInt(self) + @sizeOf(Block) + self.len;
        return @intToPtr(*Block, address);
    }

    pub fn setLen(self: *Block, len: usize) void {
        self.len = @truncate(@TypeOf(self.len), len);
    }

    comptime {
        if (@bitSizeOf(Block) != @bitSizeOf(usize)) {
            compileErrorFmt("Heap block incorrect bit size for target: ", .{@bitSizeOf(Block)});
        }
        if (@sizeOf(Block) != @sizeOf(usize)) {
            compileErrorFmt("Heap block incorrect byte size for target: ", .{@sizeOf(Block)});
        }
    }
};

const page_flags = arch.paging.PageTableEntry.generateU64(.{
    .present = true,  // Present
    .writable = true,  // Writable
    .user_accessable = false, // User accessable
    // FIXME: Turning on NX currently breaks test machine
    .no_execute = false, // No execute
});

pub var list_head: ?*Block = null;

/// Initialises an area of virtual memory for use as heap space.
///
/// The caller should ensure the page allocator is initialised and that the area has not already
/// been added as heap space.
/// The allocator will automatically map pages, so the area should be unmapped.
pub fn initHeap(heap: []u8) !void {
    if (heap.len < @sizeOf(Block)) return error.HeapTooSmall;
    const new_block_addr = std.mem.alignForward(@ptrToInt(heap.ptr), Block.alignment);
    const new_block = @intToPtr(*Block, new_block_addr);
    _ = try page_allocator.mapPage(new_block_addr, page_flags);
    new_block.* = Block.init(
        (@ptrToInt(heap.ptr) + heap.len) - new_block_addr - @sizeOf(Block),
        false,
        false,
    );
    list_head = new_block;
}

fn alloc(
    self: *Allocator,
    len: usize,
    ptr_align: u29,
    _len_align: u29,
    ret_addr: usize,
) ![]u8 {
    std.debug.assert(ptr_align != 0);
    const kernel_heap_allocator = @fieldParentPtr(KernelHeapAllocator, "allocator", self);
    const held = kernel_heap_allocator.lock.acquire();
    defer held.release();
    // Scan through list to find free space large enough
    var cur_block_maybe: ?*Block = list_head;
    while (cur_block_maybe) |cur_block| : (cur_block_maybe = cur_block.getNext()) {
        if (cur_block.used) continue;
        const unaligned_start_addr = @ptrToInt(cur_block.getStartPtr());
        const start_addr = std.mem.alignForward(unaligned_start_addr, ptr_align);
        const max_addr = unaligned_start_addr +% cur_block.len -% 1;
        const end_addr = start_addr +% len -% 1;
        if (end_addr > max_addr) continue;
        // We've found the block, reserve
        cur_block.used = true;
        // If enough space, split block into used block and free block, otherwise keep block as is
        const new_block = @intToPtr(
            *Block,
            std.mem.alignForward(end_addr + 1, Block.alignment),
        );
        const new_space_start = @ptrCast([*]u8, new_block) + @sizeOf(Block);
        if (@ptrToInt(new_space_start) < max_addr) {
            cur_block.setLen(@ptrToInt(new_block) - unaligned_start_addr);
            _ = try page_allocator.mapPage(@ptrToInt(new_block), page_flags);
            new_block.* = Block.init(
                max_addr - @ptrToInt(new_space_start) + 1,
                false,
                cur_block.next,
            );
            cur_block.next = true;
        }
        // Allocate pages
        {
            const end_page = std.mem.alignBackward(end_addr, 4096);
            var current_page = std.mem.alignBackward(unaligned_start_addr, 4096);
            while (current_page <= end_page) : (current_page += 4096) {
                _ = try page_allocator.mapPage(current_page, page_flags);
            }
        }
        return @intToPtr([*]u8, start_addr)[0 .. end_addr - start_addr + 1];
    } else {
        logger.debug("Out of blocks!", .{});
        return error.OutOfMemory;
    }
}

fn resize(
    self: *Allocator,
    buf: []u8,
    buf_align: u29,
    new_len: usize,
    _len_align: u29,
    ret_addr: usize,
) Error!usize {
    const kernel_heap_allocator = @fieldParentPtr(KernelHeapAllocator, "allocator", self);
    const held = kernel_heap_allocator.lock.acquire();
    defer held.release();
    // In-place size increasing not supported
    if (new_len > buf.len)
        return error.OutOfMemory;
    if (new_len == 0 and buf.len != 0) {
        // Free memory
        const search_addr = @ptrToInt(buf.ptr);
        var prev_block_maybe: ?*Block = null;
        var cur_block_maybe: ?*Block = list_head;
        while (cur_block_maybe) |cur_block| : ({
            prev_block_maybe = cur_block_maybe;
            cur_block_maybe = cur_block.getNext();
        }) {
            if (builtin.mode != .Debug) {
                // In debug mode we want to catch double frees
                if (!cur_block.used) continue;
            }
            const min_addr = @ptrToInt(cur_block.getStartPtr());
            const max_addr = min_addr +% cur_block.len -% 1;
            // Check if block contains allocation
            if (min_addr <= search_addr and search_addr <= max_addr) {
                if (builtin.mode == .Debug) {
                    if (!cur_block.used) {
                        @panic("allocator detected double free");
                    }
                }
                cur_block.used = false;
                // Free middle pages
                {
                    const start_page = std.mem.alignForward(min_addr, 4096);
                    const end_page = std.mem.alignBackward(max_addr, 4096);
                    var current_page = start_page;
                    while (current_page < end_page) : (current_page += 4096) {
                        _ = page_allocator.unmapPage(current_page);
                    }
                }
                // Merge forward if next block is free
                if (cur_block.getNext()) |next_block| blk: {
                    if (next_block.used) break :blk;
                    const cur_block_page = alignToPage(@ptrToInt(cur_block));
                    const next_block_page = alignToPage(@ptrToInt(next_block));
                    cur_block.len += @sizeOf(Block) + next_block.len;
                    cur_block.next = next_block.next;
                    // Check if merged block header page can be freed
                    if (cur_block_page != next_block_page) {
                        if (next_block.getNext()) |next_next_block| {
                            if (alignToPage(@ptrToInt(next_next_block)) != next_block_page) {
                                _ = page_allocator.unmapPage(next_block_page);
                            }
                        }
                    }
                }
                // Merge backward if previous block is free
                if (prev_block_maybe) |prev_block| blk: {
                    if (prev_block.used) break :blk;
                    const cur_block_page = alignToPage(@ptrToInt(cur_block));
                    const prev_block_page = alignToPage(@ptrToInt(prev_block));
                    prev_block.len += @sizeOf(Block) + cur_block.len;
                    prev_block.next = cur_block.next;
                    cur_block.* = undefined;
                    // Check if merged block header page can be freed
                    if (cur_block_page != prev_block_page) {
                        if (cur_block.getNext()) |next_block| {
                            if (alignToPage(@ptrToInt(next_block)) != cur_block_page) {
                                _ = page_allocator.unmapPage(cur_block_page);
                            }
                        }
                    }
                }
                return 0;
            }
        }
        @panic("attempt to free pointer not part of heap");
    } else {
        return new_len;
    }
}

const KernelHeapAllocator = struct {
    allocator: Allocator,
    lock: smp.SpinLock,
};

pub const heap_allocator_ptr = &heap_allocator.allocator;

var heap_allocator = KernelHeapAllocator{
    .allocator = Allocator{
        .allocFn = alloc,
        .resizeFn = resize,
    },
    .lock = smp.SpinLock.init(),
};

File: kernel_rust/src/heap.rs
use crate::arch::page_allocation;
use crate::arch::paging::{align_to_page, PageTableEntry, PAGE_SIZE};
use bitfield::bitfield;
use core::alloc::{GlobalAlloc, Layout};
use core::iter::Iterator;
use core::mem::{align_of, size_of};
use core::ptr::{self, NonNull};
use spin::Mutex;

#[cfg(target_pointer_width = "64")]
bitfield! {
    #[repr(transparent)]
    struct Block(u64);
    len_internal, set_len_internal: 61, 0;
    pub used, set_used: 62;
    pub has_next, set_has_next: 63;
}

impl Block {
    #[cfg(target_pointer_width = "64")]
    const LEN_MASK: u64 = 0x3FFF_FFFF_FFFF_FFFF;
    #[cfg(target_pointer_width = "32")]
    const LEN_MASK: u32 = 0x3FFF_FFFF;

    #[cfg(target_pointer_width = "64")]
    pub fn new(len: usize, used: bool, has_next: bool) -> Self {
        Self(len as u64 & Self::LEN_MASK | (used as u64) << 62 | (has_next as u64) << 63)
    }

    #[cfg(target_pointer_width = "32")]
    pub fn new(len: usize, used: bool, has_next: bool) -> Self {
        Self(len as u32 & Self::LEN_MASK | (used as u32) << 30 | (has_next as u32) << 31)
    }

    pub fn len(&self) -> usize {
        self.len_internal() as usize
    }

    pub fn set_len(&mut self, value: usize) {
        #[cfg(target_pointer_width = "64")]
        self.set_len_internal(value as u64);
        #[cfg(target_pointer_width = "32")]
        self.set_len_internal(value as u32);
    }

    /// Returns the start address of the inner block.
    pub fn start_address(&self) -> usize {
        self as *const Self as usize + size_of::<Self>()
    }

    pub unsafe fn get_next(&self) -> Option<NonNull<Self>> {
        if !self.has_next() {
            return None;
        }
        let address = self as *const Self as usize + size_of::<Self>() + self.len();
        Some(NonNull::new(address as *mut Self).unwrap_unchecked())
    }

    pub unsafe fn iter_mut(&mut self) -> BlockIterator {
        BlockIterator {
            current_block: Some(NonNull::from(self)),
        }
    }
}

struct BlockIterator {
    current_block: Option<NonNull<Block>>,
}

impl Iterator for BlockIterator {
    type Item = NonNull<Block>;

    fn next(&mut self) -> Option<Self::Item> {
        unsafe {
            let current_block = self.current_block?;
            self.current_block = current_block.as_ref().get_next();
            Some(current_block)
        }
    }
}

const PAGE_FLAGS: PageTableEntry = PageTableEntry::READ_WRITE;

struct KernelHeapAllocator {
    pub list_head: Mutex<Option<NonNull<Block>>>,
}

unsafe impl Sync for KernelHeapAllocator {}

unsafe impl GlobalAlloc for KernelHeapAllocator {
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        let maybe_list_head_lock = self.list_head.lock();
        let Some(list_head) = maybe_list_head_lock.map(|mut ptr| ptr.as_mut()) else {
            return ptr::null_mut();
        };
        // Scan through list to find free space large enough
        for mut current_block_ptr in list_head.iter_mut() {
            let current_block = current_block_ptr.as_mut();
            if current_block.used() {
                continue;
            }
            let unaligned_start_addr = current_block.start_address();
            let start_addr = unaligned_start_addr.next_multiple_of(layout.align());
            let max_addr = unaligned_start_addr + (current_block.len() - 1);
            let end_addr = start_addr + (layout.size() - 1);
            if end_addr > max_addr {
                continue;
            }
            // Found a suitable block, reserve
            current_block.set_used(true);
            // If enough space, split block into used and free blocks, otherwise keep block as is
            let new_block_addr = (end_addr + 1).next_multiple_of(align_of::<Block>());
            let new_space_start = new_block_addr + size_of::<Block>();
            if new_space_start < max_addr {
                current_block.set_len(new_block_addr - unaligned_start_addr);
                match page_allocation::map_page(new_block_addr, PAGE_FLAGS) {
                    Ok(_) => {}
                    Err(page_allocation::MapPageError::PageAlreadyExists) => {}
                    err @ Err(_) => err.unwrap(),
                }
                *(new_block_addr as *mut Block) = Block::new(
                    max_addr - new_space_start + 1,
                    false,
                    current_block.has_next(),
                );
                current_block.set_has_next(true);
            }
            // Allocate pages
            {
                let start_page = align_to_page(unaligned_start_addr);
                let end_page = align_to_page(end_addr);
                for page in (start_page..=end_page).step_by(PAGE_SIZE) {
                    // FIXME: `PageAlreadyExists` - check against old Zig code
                    match page_allocation::map_page(page, PAGE_FLAGS) {
                        Ok(_) => {}
                        Err(page_allocation::MapPageError::PageAlreadyExists) => {}
                        err @ Err(_) => err.unwrap(),
                    }
                }
            }
            return start_addr as *mut u8;
        }
        // Space not found, return failure
        ptr::null_mut()
    }

    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
        let search_addr = ptr as usize;
        let list_head = self.list_head.lock().unwrap().as_mut();
        let mut maybe_previous_block_ptr: Option<NonNull<Block>> = None;
        for mut current_block_ptr in list_head.iter_mut() {
            let current_block = current_block_ptr.as_mut();
            let min_addr = current_block.start_address();
            let max_addr = min_addr + (current_block.len() - 1);
            // Check if block contains allocation
            if min_addr <= search_addr && search_addr <= max_addr {
                // Check for double free in debug mode
                debug_assert!(current_block.used());
                current_block.set_used(false);
                // Free middle pages
                {
                    let start_page = min_addr.next_multiple_of(PAGE_SIZE);
                    let end_page = align_to_page(max_addr);
                    for page in (start_page..end_page).step_by(PAGE_SIZE) {
                        page_allocation::free_page(page);
                    }
                }
                let current_block_page = align_to_page(current_block as *mut Block as usize);
                // Merge forward if next block is free
                match current_block.get_next().map(|mut ptr| ptr.as_mut()) {
                    Some(next_block) if !next_block.used() => 'blk: {
                        let next_block_page = align_to_page(next_block as *mut Block as usize);
                        current_block
                            .set_len(current_block.len() + size_of::<Block>() + next_block.len());
                        current_block.set_has_next(next_block.has_next());
                        // Check if merged block header page can be freed
                        if current_block_page != next_block_page {
                            break 'blk;
                        }
                        let Some(next_next_block_ptr) = next_block.get_next() else {
                            break 'blk;
                        };
                        if align_to_page(next_next_block_ptr.as_ptr() as usize) != next_block_page {
                            page_allocation::unmap_and_free_page(next_block_page);
                        }
                    }
                    _ => {}
                }
                // Merge backward if next block is free
                match maybe_previous_block_ptr.map(|mut ptr| ptr.as_mut()) {
                    Some(previous_block) if !previous_block.used() => 'blk: {
                        if previous_block.used() {
                            break 'blk;
                        }
                        let previous_block_page =
                            align_to_page(previous_block as *mut Block as usize);
                        previous_block.set_len(
                            previous_block.len() + size_of::<Block>() + current_block.len(),
                        );
                        previous_block.set_has_next(current_block.has_next());
                        // Check if merged block header page can be freed
                        if previous_block_page != current_block_page {
                            break 'blk;
                        }
                        let Some(next_block_ptr) = current_block.get_next() else {
                            break 'blk;
                        };
                        if align_to_page(next_block_ptr.as_ptr() as usize) != current_block_page {
                            page_allocation::unmap_and_free_page(current_block_page);
                        }
                    }
                    _ => {}
                }
                return;
            }
            maybe_previous_block_ptr = Some(current_block_ptr);
        }
    }
}

#[global_allocator]
static ALLOCATOR: KernelHeapAllocator = KernelHeapAllocator {
    list_head: Mutex::new(None),
};

/// Initialises an area of virtual memory for use as heap space. The allocator will automatically
/// map pages, so the area should be unmapped.
///
/// # Safety
/// The caller guarantees this function is only called once.
pub unsafe fn init_heap(start_address: usize, length: usize) {
    let new_block_addr = start_address.next_multiple_of(align_of::<Block>());
    page_allocation::map_page(new_block_addr, PAGE_FLAGS).unwrap();
    let new_block_ptr = new_block_addr as *mut Block;
    new_block_ptr.write(Block::new(
        (start_address + length) - new_block_addr - size_of::<Block>(),
        false,
        false,
    ));
    *ALLOCATOR.list_head.lock() = Some(NonNull::new_unchecked(new_block_ptr));
}
